### PR TITLE
Log time.Duration _twice_ in all modes, as string and as nanoseconds

### LINKED ISSUE #6931: Unformatted "took" fields in JSON logs
## What

It appears as an `int` field, in nanoseconds.  For humans it is easier to read longer durations if they are formatted.  Here's [an example](https://go.dev/play/p/qWGE52sZ1pM).

Log all Durations in both formats, making it clear that the numerical format is nanoseconds.  For `"took"` these can be:

* `"took"` - formatted time, 1m20s
* `"took_nsecs"`

## Why does this happen?

* logrus uses encoding/json.
* encoding/json allows types to implement appropriate interfaces to control how they are marshaled: Types can implement the MarshalJSON method or MarshalText
  - time.Time implements both, time.Duration implements neither
  - It is known not to work, but there are [no current plans to fix it](https://github.com/golang/go/issues/4712) for the past 5-ish years.
  - So json.Marshal sees it's an int and encodes its raw value; time.Duration is [a count of nanoseconds](https://pkg.go.dev/time#Duration).
* There is no facility in encoding/json to change the encoding of a type except on that type
* There is no facility in logrus to change the JSON encoding of a type.

However our logging package wraps logrus, and we can change WithField and WithFields to fix this.

### PR DESCRIPTION
**Pull request by arielshaqed**
Append "_str" or "_nsecs" to the label to be absolutely clear what is being logged.

Fixes #6931.

### CODE REVIEW

**Comment by arielshaqed** [comment]
As an alternative we might consider using `label` always to hold nanoseconds (currently this varies between text and JSON logs) and adding an additional `label_str` value that always holds strings.

File: pkg/logging/logger.go [hunk 1]
```diff
@@ -98,8 +111,19 @@ func SetLevel(level string) {
...
+			return fmt.Errorf("close log writer: %w", err)
+		}
+	}
+	openLoggers = nil
+	return nil
+}
+
 func SetOutputs(outputs []string, fileMaxSizeMB, filesKeep int) {
 	var writers []io.Writer
+	CloseWriters()
```
Comment (idanovo) [inline pkg/logging/logger.go:126]: Please handle the returned error

File: pkg/logging/logger_test.go [hunk 2]
```diff
@@ -58,3 +61,61 @@ func TestSetOutputs(t *testing.T) {
...
+			if err != nil {
+				t.Fatalf("Open %s to read contents: %s", log, err)
+			}
+			contents, err := io.ReadAll(r)
+			if err != nil {
+				t.Fatalf("Read %s contents: %s", log, err)
+			}
+
+			for _, expected := range tc.FormatDurations("xyzzy", duration) {
+				if !bytes.Contains(contents, []byte(expected)) {
```
Comment (idanovo) [inline pkg/logging/logger_test.go:115]: I'm not sure I understand from this test what do you expect to see in the logs

**Review by idanovo — CHANGES_REQUESTED** [review]
Liked the change even though the change in the `WithField` function is a little bit cumbersome IMO.
Few comments

File: pkg/logging/logger.go (hunk 1 above)
Comment (arielshaqed) [reply pkg/logging/logger.go:126]: Done, thanks!

File: pkg/logging/logger_test.go (hunk 2 above)
Comment (arielshaqed) [reply pkg/logging/logger_test.go:115]: I expect to see in text:
```
...xyzzy_str=17h19m...xyzzy_nsecs=9999473948...
```
and in JSON:
```json
{
  "lots of":"stuff removed",
  "xyzzy_str":"17h19m",
  "xyzzy_nsecs":9999473948
}
```
but of course with some other value for xyzzy_nsecs.

**Review by arielshaqed — COMMENTED** [review]
Thanks!  PTAL.

**Review by idanovo — DISMISSED** [review]
LGTM, thanks

**Comment by arielshaqed** [comment]
Also fixes #3744

**Comment by arielshaqed** [comment]
Oh sorry, your comment about WithField made me realize I hadn't fixed WithFields!

**Comment by arielshaqed** [comment]
@idanovo sorry, PTAL.

**Review by idanovo — APPROVED** [review]
LGTM

**Comment by arielshaqed** [comment]
Thanks!

One last bug (nil values) and I'm pulling.

**Review by nopcoder — DISMISSED** [review]
Running the following

```go
package main

import (
	"time"

	"github.com/sirupsen/logrus"
)

func main() {
	log := logrus.New()
	passed := time.Now().Add(-time.Second * 5)
	since := time.Since(passed)
	log.WithField("since", since).Info("hello world")
}
```

Output
```
INFO[0000] hello world                                   since=5.000026375s
```

So, I think we are doing something wrong, or too complex or we are using an old version of logrus.

**Comment by arielshaqed** [comment]
@nopcoder says:

> So, I think we are doing something wrong, or too complex or we are using an old version of logrus.

See the issue for the underlying behaviour.  JSON logs work differently from _real_ logs, see [this playground](https://go.dev/play/p/qWGE52sZ1pM).

**Comment by nopcoder** [comment]
> @nopcoder says:
> 
> > So, I think we are doing something wrong, or too complex or we are using an old version of logrus.
> 
> See the issue for the underlying behaviour. JSON logs work differently from these _text_ logs, see [this playground](https://go.dev/play/p/qWGE52sZ1pM).

Thanks for pointing out the JSON part. I still think this is a very specific feature to add to the logger, which would replace my keys to specify the output format. I'm not sure if it would be easier for users to pass the duration as a string or to discover this underlying functionality.

Would adding `.WithDuration("prefix", value)` method clarify the code and reduce the runtime cost of operation of the logger in this case?"

**Comment by arielshaqed** [comment]
> > @nopcoder says:
> > > So, I think we are doing something wrong, or too complex or we are using an old version of logrus.
> > 
> > 
> > See the issue for the underlying behaviour. JSON logs work differently from these _text_ logs, see [this playground](https://go.dev/play/p/qWGE52sZ1pM).
> 
> Thanks for pointing out the JSON part. I still think this is a very specific feature to add to the logger, which would replace my keys to specify the output format. I'm not sure if it would be easier for users to pass the duration as a string or to discover this underlying functionality.

As an alternative: **we could decide to keep on using "duration" as one of the keys, and just have a prefix on the other.**  For instance for `.WithField("key", d)` we could have `d.String()` as `"key"` and `d.Nanoseconds()` as `"key_nsecs"`.  Then there is no negative surprise from reading the code: I expected to see `"key"` in the logs, and sure enough - it's there.  There is only a positive surprise: I might be happy to see `"key_nsecs"` in the code, and investigate from there.

> Would adding `.WithDuration("prefix", value)` method clarify the code and reduce the runtime cost of operation of the logger in this case?"

I don't think so.

1. Nobody will use it, because `.WithField("key", value)` just _works_.
1. It doesn't replace `.WithFields()`, which would make everything a bit more awkward.  For instance [this code](https://github.com/treeverse/lakeFS/blob/1750443dda6bbf44be6d63fa2cca79d6b692b440/pkg/httputil/logging.go#L89-L96) would need to become:
   ```go
	loggingFields := logging.Fields{
		"status_code":    writer.StatusCode,
		"sent_bytes":     writer.ResponseSize,
		"client":         client,
		logging.LogAudit: true,
		"source_ip":      sourceIP,
	}
	log := logging.
		FromContext(r.Context()).
		WithDuration("took", time.Since(startTime).
		WithFields(loggingFields)
   ```
   which is a bit naff.


WDYT about the "alternative" above?

**Comment by nopcoder** [comment]
> WDYT about the "alternative" above?

Like the alternative, but can we keep the original key with the original behaviour as nsec and the additional key will be the formatted one?

**Review by nopcoder — APPROVED** [review]
Like the alternative, but can we keep the original key with the original behaviour as nsec and the additional key will be the formatted one?
Will not request the PR if it will not be the case. I totally agree that people will forgot.
Just usually I prefer simple solutions that doesn't cost in runtime reflect of the values we pass to the logger and by having a simple logging it enables simple integration or switch to other implementations.

**Comment by arielshaqed** [comment]
> Like the alternative, but can we keep the original key with the original behaviour as nsec and the additional key will be the formatted one? Will not request the PR if it will not be the case. I totally agree that people will forgot. Just usually I prefer simple solutions that doesn't cost in runtime reflect of the values we pass to the logger and by having a simple logging it enables simple integration or switch to other implementations.

Sounds like a plan! That way any existing code that processes JSON logs will continue to work.

**Pull request merged**